Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -582,6 +583,14 @@ extension SourceKitServer {
         self.dynamicallyRegisterCapability($0, registry)
       }
     }
+    /// Request the client to send us DidChangeWatchedFileNotifications for all new files.
+    /// Currently, these notifications are only handled by SwiftPMWorkspace, which will reload the package when a new file is added.
+    let watchers = [
+      FileSystemWatcher(globPattern: "**", kind: [.create])
```

Comment:
How about
* Also handle `.delete` on all files
* Add a pattern for `Package.swift` (or maybe `**/Package.swift`?) with `.change`

Reply:
Should this be in SourceKitServer though or should we let the BuildSystem tell us what we should watch?

Reply:
I'm fine with doing it either way in this PR.  You're right that in principle there should be interaction with the build system for this.  I'm not sure what that API should actually look like though, because in principle the build system also wants to watch paths that might be outside the workspace -- e.g. the build directory may not be in the workspace -- and couldn't necessarily rely on the LSP notifications for those specific paths. So I would say do the simple thing and make it a fixed list either here or as a property on the build system :shrug:

Reply:
> * Also handle `.delete` on all files

I thought about the same but don’t think it’s necessary. There are no new build settings to complete and the old ones, which reference the deleted file, still seem to work for sourcekitd.

> * Add a pattern for `Package.swift` (or maybe `**/Package.swift`?) with `.change`

I’d like to check how many notifications VSCode sends us if listen to change events as well. If we receive a change notification for every keystroke and all keystrokes trigger a complete package reload, this might be too performance critical. Let’s do this in a follow-up PR.

> Should this be in SourceKitServer though or should we let the BuildSystem tell us what we should watch?

Since we are not really narrowing down on any specific pattern here, I don’t think there’s a point in letting the `BuildSystem` tell us what to watch for, it could just ignore the events it doesn’t care for. I agree that it should be a `BuildSystem` property if we only listened for changes to `.swift` files, but let’s leave this as an optimization for follow-up PRs, I’d say.

Reply:
> I’d like to check how many notifications VSCode sends us if listen to change events as well. If we receive a change notification for every keystroke and all keystrokes trigger a complete package reload, this might be too performance critical. Let’s do this in a follow-up PR.

It would not be correct to send changes per keystroke, these APIs are only for changes to the file, not to the in-memory text buffer.  We can do it in a follow-up PR if you like, but I think this is important before we close the SR issue.

Reply:
Or for the SR, you could un-dupe https://bugs.swift.org/browse/SR-15679. That would be fine too.

Reply:
RE: delete

It's great that sourcekit is being lenient about missing files, but I think we should still strive to use accurate settings.
* clangd: We use the index to choose a main file for a header, and if that main file is removed, we need to update our settings to use a different main file if possible.  This should already work in the build system manager except that it needs to get the `.removedOrUnavailable` update from the underlying build system.
* swiftpm: We're implicitly relying on swiftpm to not make any *other* changes to settings based on the removal of a file except removing it from the module sources. That may be true today, but it could change.
* other uses of settings: We shouldn't assume we won't use these settings for other things like invoking the compiler in the future, and it will fail if we're missing an input.

Reply:
Sounds good. I’ve added `.delete`.